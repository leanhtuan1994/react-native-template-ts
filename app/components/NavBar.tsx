import React from "react";
import { StyleSheet } from "react-native";
import { Header, HeaderProps } from "react-native-elements";
import { Colors, FontSize, FontWeight, GlobalStyles } from "styles";
import { useNavigation, StackActions } from "@react-navigation/native";
import { useBackHandler } from "@react-native-community/hooks";
import { isIOS13 } from "utils/platform";

type Props = {
  type?: "modal" | "stack" | "none";
  title?: string;
  backToTop?: boolean;
  onGoBack?: () => void;
};

const titleStyle = GlobalStyles.font({
  fontSize: FontSize.Header,
  fontWeight: FontWeight.SemiBold,
  color: Colors.Black,
});

const NavBar: React.FC<HeaderProps & Props> = ({
  type = "stack",
  title,
  backToTop,
  containerStyle,
  onGoBack,
  ...props
}) => {
  const navigation = useNavigation();

  const isSpecialHeader = isIOS13 && type === "modal";

  const popToTop = () => {
    navigation.dispatch(StackActions.popToTop());
  };

  useBackHandler(() => {
    if (navigation.canGoBack()) {
      if (backToTop) {
        popToTop();
        return true;
      }
      onGoBack && onGoBack();
      navigation.goBack();
    }
    return true;
  });

  return (
    <Header
      statusBarProps={
        isSpecialHeader
          ? { barStyle: "light-content" }
          : {
              barStyle: "dark-content",
              translucent: true,
              backgroundColor: Colors.White,
              networkActivityIndicatorVisible: true,
            }
      }
      containerStyle={StyleSheet.flatten([
        isSpecialHeader
          ? {
              paddingTop: 0,
              height: 64,
            }
          : {},
        containerStyle,
      ])}
      leftComponent={
        type === "none"
          ? (title && { text: title, style: titleStyle }) || undefined
          : {
              type: "feather",
              icon: type === "stack" ? "arrow-left" : "x",
              onPress: () => {
                if (navigation.canGoBack()) {
                  if (backToTop) {
                    popToTop();
                  }
                  onGoBack && onGoBack();
                  navigation.goBack();
                }
              },
              size: 24,
              hitSlop: { top: 6, bottom: 6, left: 6, right: 6 },
            }
      }
      centerComponent={
        type !== "none" && title
          ? { text: title, style: titleStyle }
          : undefined
      }
      {...props}
    />
  );
};

export default NavBar;
