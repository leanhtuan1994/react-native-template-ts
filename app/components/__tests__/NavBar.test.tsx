import React from "react";
import { shallow } from "enzyme";

import NavBar from "../NavBar";

describe("NavBar", () => {
  it("navbar match to snapshoot", () => {
    const nav = shallow(<NavBar />);
    expect(nav).toMatchSnapshot();
  });
});
