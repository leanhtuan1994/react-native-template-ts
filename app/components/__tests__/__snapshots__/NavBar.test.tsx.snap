// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar navbar match to snapshoot 1`] = `
<ForwardRef(Themed.Header)
  containerStyle={Object {}}
  leftComponent={
    Object {
      "hitSlop": Object {
        "bottom": 6,
        "left": 6,
        "right": 6,
        "top": 6,
      },
      "icon": "arrow-left",
      "onPress": [Function],
      "size": 24,
      "type": "feather",
    }
  }
  statusBarProps={
    Object {
      "backgroundColor": "#FFFFFF",
      "barStyle": "dark-content",
      "networkActivityIndicatorVisible": true,
      "translucent": true,
    }
  }
/>
`;
